Allow deleting packages if only used by project.json projects

TimBarham opened the pull request:
When uninstalling a package from a `packages.config` project, and checking if we can delete it from the packages folder, we don't care if that package is being used by a `project.json` project.

I encountered this while trying to get package deletion working properly for the project upgrade feature.

Note I've also created extracted some code into a method `DeletePackageDirectoriesAsync` as I want to be able to re-use that logic elsewhere.

TimBarham committed:
Allow deleting packages if only used by project.json projects.

When uninstalling a package from a packages.config project, and checking if we can delete it from the packages folder, we don't care if that package is being used by a project.json project.

This resolves NuGet/Home#2990

TimBarham commented:
See NuGet/Home#2990.

yishaigalatzer commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1841,6 +1818,38 @@ await PreviewBuildIntegratedProjectActionsAsync(buildIntegratedProject, actions,
+        /// Deletes the package directories for the supplied list of packages. If an error occurs (for example,
+        /// a directory is in use), marks the directory for deletion when VS is restarted.
+        /// </summary>
+        public async Task DeletePackageDirectoriesAsync(INuGetProjectContext nuGetProjectContext, HashSet<PackageIdentity> packages, CancellationToken token)
```
break long line

yishaigalatzer replied:
Don't use `HashSet` on a public API, this could be an `IEnumerable` as far as the method is concerned

yishaigalatzer commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2281,7 +2290,7 @@ public bool PackageExistsInPackagesFolder(PackageIdentity packageIdentity)
+                && !await PackageExistsInAnotherNuGetProject(nuGetProject, packageIdentity, SolutionManager, token, excludeIntegrated: true))
             {
                 packageWithDirectoriesToBeDeleted.Add(packageIdentity);
             }
```
remove TODO and commented out code below

yishaigalatzer commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -2298,7 +2307,7 @@ public bool PackageExistsInPackagesFolder(PackageIdentity packageIdentity)
         /// other projects in the solution.
         /// </summary>
-        public static async Task<bool> PackageExistsInAnotherNuGetProject(NuGetProject nuGetProject, PackageIdentity packageIdentity, ISolutionManager solutionManager, CancellationToken token)
+        public static async Task<bool> PackageExistsInAnotherNuGetProject(NuGetProject nuGetProject, PackageIdentity packageIdentity, ISolutionManager solutionManager, CancellationToken token, bool excludeIntegrated = false)
```
avoid default parameters on public APIs, and break long line

yishaigalatzer replied:
the `excludeIntegrated` parameter is not very clear to me. will it remove both xproj and csproj + project.json or just one of the two? It should remove both.

TimBarham replied:
OK, I've switched to an overload. Regarding that parameter - yes, both `xproj` and `csproj` with `project.json` will be ignored if that parameter is specified. I've renamed it `ignoreNuGetIntegratedProjects` (which relates to the fact that the project types in question implement `INuGetIntegratedProject`) - is that sufficient, or do you have any other suggestions?

yishaigalatzer commented:
:watch:

TimBarham committed:
Changes in reponse to PR feedback.

Also, moved DeletePackageDirectoriesAsync back inline as I no longer need this functionality elsewhere.

TimBarham commented:
Made changes in reponse to PR feedback. Also, moved `DeletePackageDirectoriesAsync()` back inline as I no longer need this functionality elsewhere.

TimBarham commented:
Ping.

emgarten commented:
@rrelyea can you take a look at https://github.com/NuGet/Home/issues/2990 and determine if this should go into 4.1.0 or 4.3.0?  If not we should close this PR.

emgarten commented:
Moving this issue to the backlog. At this point this PR is very out of date and the work would need to be re-done.
